fifo/consumer.go

```diff
@@ -18,129 +19,244 @@ import (
+	}
+}
+
+func (c *Consumer) Read(ctx context.Context, thresh int, timeout time.Duration) (zbuf.Array, error) {
+	var batch zbuf.Array
+	var size int
+	events := c.consumer.Events()
+	for {
+		select {
+		case ev := <-events:
+			if ev == nil {
+				// channel closed
+				return batch, nil
 			}
-			bytes, err := zavro.Decode(msg.Value[5:], avroTypeRecord)
+			rec, err := c.handle(ev)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			var rec *zng.Record
-			if c.wrap {
-				rec, err = c.wrapRecord(zctx, metaType, recType, bytes, msg.TopicPartition)
-				if err != nil {
-					return err
-				}
-			} else {
-				rec = zng.NewRecord(recType, bytes)
+			if rec == nil {
+				// unknown event
+				continue
 			}
-			if err := w.Write(rec); err != nil {
-				return err
-			}
-			if (msg.TopicPartition.Offset + 1) >= c.highWater {
-				break
+			batch.Append(rec)
+			size += len(rec.Bytes)
+			if size > thresh {
+				return batch, nil
 			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(timeout):
+			return batch, nil
 		}
 	}
-	return nil
 }
 
-func (c *Consumer) wrapRecord(zctx *zson.Context, metaType, typ zng.Type, bytes []byte, meta kafka.TopicPartition) (*zng.Record, error) {
-	outerType, ok := c.types[typ]
-	if !ok {
-		cols := []zng.Column{
-			{"kafka", metaType},
-			//{"key", keyType}, XXX not yet
-			{"value", typ},
+func (c *Consumer) handle(ev kafka.Event) (*zng.Record, error) {
+	switch ev := ev.(type) {
+	case kafka.Error:
+		return nil, ev
+	case *kafka.Message:
+		key, err := c.decodeAvro(ev.Key)
+		if err != nil {
+			return nil, err
 		}
-		var err error
-		outerType, err = zctx.LookupTypeRecord(cols)
+		val, err := c.decodeAvro(ev.Value)
 		if err != nil {
 			return nil, err
 		}
-		c.types[typ] = outerType
+		return c.wrapRecord(key, val, ev.TopicPartition)
+	default:
+		return nil, nil
+	}
+}
+
+func (c *Consumer) wrapRecord(key, val zng.Value, meta kafka.TopicPartition) (*zng.Record, error) {
+	outerType, err := c.outerType(key.Type, val.Type)
+	if err != nil {
+		return nil, err
 	}
 	var b zcode.Builder
-	// {topic:string,partition:int64,offset:int64}
-	b.BeginContainer()
-	b.AppendPrimitive([]byte(*meta.Topic))
-	b.AppendPrimitive(zng.EncodeInt(int64(meta.Partition)))
-	b.AppendPrimitive(zng.EncodeInt(int64(meta.Offset)))
-	b.EndContainer()
-	b.AppendContainer(bytes)
+	if c.metaType != nil {
+		// kafka:{topic:string,partition:int64,offset:int64}
+		b.BeginContainer()
+		b.AppendPrimitive([]byte(*meta.Topic))
+		b.AppendPrimitive(zng.EncodeInt(int64(meta.Partition)))
+		b.AppendPrimitive(zng.EncodeInt(int64(meta.Offset)))
+		b.EndContainer()
+	}
+	b.AppendContainer(key.Bytes)
+	b.AppendContainer(val.Bytes)
 	return zng.NewRecord(outerType, b.Bytes()), nil
 }
+
+func (c *Consumer) decodeAvro(b []byte) (zng.Value, error) {
+	if len(b) == 0 {
+		return zng.Value{Type: zng.TypeNull}, nil
+	}
+	if len(b) < 5 {
+		return zng.Value{}, fmt.Errorf("bad kafka-avro value in topic: len %d", len(b))
```

Comment:
```suggestion
		return zng.Value{}, fmt.Errorf("bad Avro value in Kafka topic: len %d", len(b))
```

Reply:
But this really would be the Kafka-Avro framing error not necessarily a problem with the Avro itself.  (The Kafka people invented the framing of putting an unused byte plus a 4-byte integer schema ID at the front of the encoded Avro message.) I will clarify.